File: src/lib/content/ai-slop-detector.ts
/**
 * AI Slop Detection for Generated Content
 * Detects AI-generated patterns and quality issues
 * 
 * Based on research: Google E-E-A-T framework and AI content detection signals
 */

export interface SlopDetectionResult {
  aiProbability: number; // 0-1 (0=human, 1=AI)
  qualityScore: number; // 0-10 (10=best)
  flags: string[];
  recommendations: string[];
  metrics: {
    wordCount: number;
    slopCount: number;
    slopPhrases: Array<{ phrase: string; count: number }>;
    emDashCount: number;
    emDashRatio: number;
    sentenceCount: number;
    avgSentenceLength: number;
    sentenceStdDev: number;
    hedgeCount: number;
    hedgeRatio: number;
    vagueCount: number;
    personalCount: number;
    hasCode: boolean;
    hasNumbers: boolean;
    hasLinks: boolean;
  };
}

const SLOP_PHRASES = [
  'delve', 'delve into', 'delving',
  'leverage', 'leveraging',
  'utilize', 'utilization',
  'robust', 'seamless', 'cutting-edge', 'state-of-the-art',
  'revolutionary', 'game-changing', 'transformative',
  "it's important to note", 'it should be noted',
  "in today's fast-paced world", 'in this day and age',
  'at the end of the day', 'first and foremost',
  'it goes without saying', 'needless to say',
];

const HEDGE_WORDS = [
  'may', 'might', 'could', 'possibly', 'generally',
  'typically', 'often', 'usually', 'in most cases',
];

const VAGUE_PHRASES = [
  'many experts', 'studies show', 'research indicates',
  "it's widely known", "it's well established", "it's commonly accepted",
];

const PERSONAL_MARKERS = [
  'i tried', 'i tested', 'in my experience', 'i found',
  'we built', 'we discovered', 'our team', 'i noticed',
  'after testing', 'when i used',
];

export function detectAISlop(text: string): SlopDetectionResult {
  const flags: string[] = [];
  const recommendations: string[] = [];

  const textLower = text.toLowerCase();
  const words = text.split(/\s+/);
  const wordCount = words.length;

  // 1. AI Slop Phrases
  let slopCount = 0;
  const slopFound: Array<{ phrase: string; count: number }> = [];

  for (const phrase of SLOP_PHRASES) {
    const count = (textLower.match(new RegExp(phrase, 'g')) || []).length;
    if (count > 0) {
      slopCount += count;
      slopFound.push({ phrase, count });
    }
  }

  if (slopCount > 0) {
    flags.push(`AI slop phrases: ${slopCount} total`);
    recommendations.push('Remove AI slop phrases, use natural language');
  }

  // 2. Em Dash Overuse (AI loves em dashes)
  const emDashCount = (text.match(/—/g) || []).length;
  const emDashRatio = wordCount > 0 ? (emDashCount / (wordCount / 100)) : 0;

  if (emDashRatio > 1) {
    flags.push(`Overuse of em dashes (${emDashCount} total, ${emDashRatio.toFixed(1)} per 100 words)`);
    recommendations.push('Replace em dashes with periods or colons');
  }

  // 3. Sentence Length Uniformity
  const sentences = text.split(/[.!?]+/).filter(s => s.trim().length > 0);
  const sentenceLengths = sentences.map(s => s.split(/\s+/).length);

  let stdDev = 0;
  if (sentenceLengths.length > 5) {
    const avgLength = sentenceLengths.reduce((a, b) => a + b, 0) / sentenceLengths.length;
    const variance = sentenceLengths.reduce((sum, len) => sum + Math.pow(len - avgLength, 2), 0) / sentenceLengths.length;
    stdDev = Math.sqrt(variance);

    // AI text has low variance (uniform sentences)
    if (stdDev < 5) {
      flags.push(`Uniform sentence length (std dev: ${stdDev.toFixed(1)})`);
      recommendations.push('Vary sentence length (mix short punchy + long complex)');
    }
  }

  // 4. Hedging Language (AI is overly cautious)
  let hedgeCount = 0;
  for (const hedge of HEDGE_WORDS) {
    hedgeCount += (textLower.match(new RegExp(`\\b${hedge}\\b`, 'g')) || []).length;
  }
  const hedgeRatio = wordCount > 0 ? hedgeCount / wordCount : 0;

  if (hedgeRatio > 0.02) {
    flags.push(`Excessive hedging language (${(hedgeRatio * 100).toFixed(1)}%)`);
    recommendations.push('Be more direct, reduce qualifiers');
  }

  // 5. Vague Claims Without Citations
  let vagueCount = 0;
  for (const phrase of VAGUE_PHRASES) {
    vagueCount += (textLower.match(new RegExp(phrase, 'g')) || []).length;
  }

  if (vagueCount > 2) {
    flags.push(`Vague claims without citations (${vagueCount}x)`);
    recommendations.push('Add specific citations and sources');
  }

  // 6. Personal Experience Markers (good = human)
  let personalCount = 0;
  for (const marker of PERSONAL_MARKERS) {
    personalCount += (textLower.match(new RegExp(marker, 'g')) || []).length;
  }

  if (personalCount === 0) {
    flags.push('No personal experience markers');
    recommendations.push("Add personal testing/experience ('I tested...', 'We found...')");
  }

  // 7. Specific Examples (good = human)
  const hasCode = text.includes('```') || text.includes('`');
  const hasNumbers = /\d/.test(text);
  const hasLinks = /https?:\/\//.test(text) || /\[.*\]\(.*\)/.test(text);

  if (!hasCode) {
    flags.push('No code examples');
    recommendations.push('Add code examples with syntax highlighting');
  }
  if (!hasNumbers) {
    flags.push('No specific numbers/data');
    recommendations.push('Add specific metrics, measurements, or data');
  }
  if (!hasLinks) {
    flags.push('No citations/links');
    recommendations.push('Add citations to sources (Reddit, GitHub, docs)');
  }

  // 8. Generic Structures (AI patterns)
  const genericPatterns = [
    'there are several reasons why',
    "let's explore the key factors",
    'here are some important considerations',
    "in conclusion, it's clear that",
    "to summarize, we've discussed",
  ];

  let genericCount = 0;
  for (const pattern of genericPatterns) {
    genericCount += (textLower.match(new RegExp(pattern, 'g')) || []).length;
  }

  if (genericCount > 1) {
    flags.push(`Generic AI structures (${genericCount}x)`);
    recommendations.push('Use more direct, specific language');
  }

  // Calculate AI probability
  const flagWeight = flags.length * 0.08; // Each flag adds 8%
  const slopWeight = Math.min(0.3, slopCount * 0.02); // Slop phrases add up to 30%
  const uniformityWeight = stdDev < 5 ? 0.2 : 0;

  let aiProbability = Math.min(1.0, flagWeight + slopWeight + uniformityWeight);

  // Calculate quality score (inverse of AI probability)
  let qualityScore = Math.max(0, 10 - aiProbability * 10);

  // Adjust quality based on positive signals
  if (personalCount > 0) qualityScore += 0.5;
  if (hasCode) qualityScore += 0.5;
  if (hasNumbers) qualityScore += 0.3;
  if (hasLinks) qualityScore += 0.3;

  qualityScore = Math.min(10, qualityScore);

  return {
    aiProbability: Math.round(aiProbability * 100) / 100,
    qualityScore: Math.round(qualityScore * 10) / 10,
    flags,
    recommendations,
    metrics: {
      wordCount,
      slopCount,
      slopPhrases: slopFound,
      emDashCount,
      emDashRatio: Math.round(emDashRatio * 100) / 100,
      sentenceCount: sentences.length,
      avgSentenceLength: sentenceLengths.length > 0
        ? Math.round((sentenceLengths.reduce((a, b) => a + b, 0) / sentenceLengths.length) * 10) / 10
        : 0,
      sentenceStdDev: Math.round(stdDev * 10) / 10,
      hedgeCount,
      hedgeRatio: Math.round(hedgeRatio * 1000) / 1000,
      vagueCount,
      personalCount,
      hasCode,
      hasNumbers,
      hasLinks,
    },
  };
}

export function printDetectionReport(detection: SlopDetectionResult): void {
  console.log('\n' + '='.repeat(60));
  console.log('AI SLOP DETECTION REPORT');
  console.log('='.repeat(60));

  console.log(`\n📊 Overall Scores:`);
  console.log(`   AI Probability: ${(detection.aiProbability * 100).toFixed(0)}%`);
  console.log(`   Quality Score: ${detection.qualityScore}/10`);

  if (detection.flags.length > 0) {
    console.log(`\n⚠️  Flags (${detection.flags.length}):`);
    detection.flags.forEach(flag => console.log(`   - ${flag}`));
  } else {
    console.log(`\n✅ No flags detected`);
  }

  if (detection.recommendations.length > 0) {
    console.log(`\n💡 Recommendations:`);
    detection.recommendations.forEach(rec => console.log(`   - ${rec}`));
  }

  const m = detection.metrics;
  console.log(`\n📈 Metrics:`);
  console.log(`   Words: ${m.wordCount.toLocaleString()}`);
  console.log(`   Sentences: ${m.sentenceCount}`);
  console.log(`   Avg Sentence Length: ${m.avgSentenceLength} words`);
  console.log(`   Sentence Variation (std dev): ${m.sentenceStdDev}`);
  console.log(`   AI Slop Phrases: ${m.slopCount}`);
  console.log(`   Em Dashes: ${m.emDashCount} (${m.emDashRatio} per 100 words)`);
  console.log(`   Personal Markers: ${m.personalCount}`);
  console.log(`   Has Code: ${m.hasCode ? '✅' : '❌'}`);
  console.log(`   Has Numbers: ${m.hasNumbers ? '✅' : '❌'}`);
  console.log(`   Has Links: ${m.hasLinks ? '✅' : '❌'}`);

  if (m.slopPhrases.length > 0) {
    console.log(`\n🚫 AI Slop Phrases Found:`);
    m.slopPhrases.forEach(slop => console.log(`   - '${slop.phrase}' (${slop.count}x)`));
  }

  console.log('\n' + '='.repeat(60));

  // Overall verdict
  if (detection.qualityScore >= 8) {
    console.log('✅ PASS: Article meets quality standards');
  } else if (detection.qualityScore >= 6) {
    console.log('⚠️  REVIEW: Article needs minor improvements');
  } else {
    console.log('❌ REJECT: Article needs major revision or regeneration');
  }

  console.log('='.repeat(60) + '\n');
}

File: lambda/lib/ai_slop_detector.py
"""
AI Slop Detection for Generated Content
Detects AI-generated patterns and quality issues
"""

import re
from typing import Dict, List

def detect_ai_slop(text: str) -> Dict:
    """
    Detect AI-generated content patterns and quality issues
    
    Args:
        text: Article content to analyze
    
    Returns:
        {
            'ai_probability': float,  # 0-1 (0=human, 1=AI)
            'quality_score': float,   # 0-10 (10=best)
            'flags': list,            # Specific issues found
            'recommendations': list,  # How to improve
            'metrics': dict          # Detailed metrics
        }
    """
    flags = []
    recommendations = []
    
    text_lower = text.lower()
    word_count = len(text.split())
    
    # 1. AI Slop Phrases
    slop_phrases = [
        'delve', 'delve into', 'delving',
        'leverage', 'leveraging',
        'utilize', 'utilization',
        'robust', 'seamless', 'cutting-edge', 'state-of-the-art',
        'revolutionary', 'game-changing', 'transformative',
        'it\'s important to note', 'it should be noted',
        'in today\'s fast-paced world', 'in this day and age',
        'at the end of the day', 'first and foremost',
        'it goes without saying', 'needless to say'
    ]
    
    slop_count = 0
    slop_found = []
    for phrase in slop_phrases:
        count = text_lower.count(phrase)
        if count > 0:
            slop_count += count
            slop_found.append({'phrase': phrase, 'count': count})
    
    if slop_count > 0:
        flags.append(f"AI slop phrases: {slop_count} total")
        recommendations.append("Remove AI slop phrases, use natural language")
    
    # 2. Em Dash Overuse (AI loves em dashes)
    em_dash_count = text.count('—')
    em_dash_ratio = (em_dash_count / (word_count / 100)) if word_count > 0 else 0
    
    if em_dash_ratio > 1:  # More than 1 per 100 words
        flags.append(f"Overuse of em dashes ({em_dash_count} total, {em_dash_ratio:.1f} per 100 words)")
        recommendations.append("Replace em dashes with periods or colons")
    
    # 3. Sentence Length Uniformity
    sentences = [s.strip() for s in re.split(r'[.!?]+', text) if s.strip()]
    sentence_lengths = [len(s.split()) for s in sentences]
    
    std_dev = 0
    if len(sentence_lengths) > 5:
        avg_length = sum(sentence_lengths) / len(sentence_lengths)
        variance = sum((x - avg_length) ** 2 for x in sentence_lengths) / len(sentence_lengths)
        std_dev = variance ** 0.5
        
        # AI text has low variance (uniform sentences)
        if std_dev < 5:
            flags.append(f"Uniform sentence length (std dev: {std_dev:.1f})")
            recommendations.append("Vary sentence length (mix short punchy + long complex)")
    
    # 4. Hedging Language (AI is overly cautious)
    hedges = ['may', 'might', 'could', 'possibly', 'generally', 
              'typically', 'often', 'usually', 'in most cases']
    
    hedge_count = sum(text_lower.count(h) for h in hedges)
    hedge_ratio = hedge_count / word_count if word_count > 0 else 0
    
    if hedge_ratio > 0.02:  # >2% hedging
        flags.append(f"Excessive hedging language ({hedge_ratio*100:.1f}%)")
        recommendations.append("Be more direct, reduce qualifiers")
    
    # 5. Vague Claims Without Citations
    vague_phrases = [
        'many experts', 'studies show', 'research indicates',
        'it\'s widely known', 'it\'s well established', 'it\'s commonly accepted'
    ]
    
    vague_count = sum(text_lower.count(p) for p in vague_phrases)
    if vague_count > 2:
        flags.append(f"Vague claims without citations ({vague_count}x)")
        recommendations.append("Add specific citations and sources")
    
    # 6. Personal Experience Markers (good = human)
    personal_markers = [
        'i tried', 'i tested', 'in my experience', 'i found',
        'we built', 'we discovered', 'our team', 'i noticed',
        'after testing', 'when i used'
    ]
    
    personal_count = sum(text_lower.count(p) for p in personal_markers)
    if personal_count == 0:
        flags.append("No personal experience markers")
        recommendations.append("Add personal testing/experience ('I tested...', 'We found...')")
    
    # 7. Specific Examples (good = human)
    has_code = '```' in text or '`' in text
    has_numbers = any(char.isdigit() for char in text)
    has_links = 'http' in text or '[' in text and ']' in text
    
    if not has_code:
        flags.append("No code examples")
        recommendations.append("Add code examples with syntax highlighting")
    if not has_numbers:
        flags.append("No specific numbers/data")
        recommendations.append("Add specific metrics, measurements, or data")
    if not has_links:
        flags.append("No citations/links")
        recommendations.append("Add citations to sources (Reddit, GitHub, docs)")
    
    # 8. Generic Structures (AI patterns)
    generic_patterns = [
        'there are several reasons why',
        'let\'s explore the key factors',
        'here are some important considerations',
        'in conclusion, it\'s clear that',
        'to summarize, we\'ve discussed'
    ]
    
    generic_count = sum(text_lower.count(p) for p in generic_patterns)
    if generic_count > 1:
        flags.append(f"Generic AI structures ({generic_count}x)")
        recommendations.append("Use more direct, specific language")
    
    # Calculate AI probability
    # More flags = higher AI probability
    flag_weight = len(flags) * 0.08  # Each flag adds 8%
    slop_weight = min(0.3, slop_count * 0.02)  # Slop phrases add up to 30%
    uniformity_weight = 0.2 if std_dev < 5 else 0
    
    ai_probability = min(1.0, flag_weight + slop_weight + uniformity_weight)
    
    # Calculate quality score (inverse of AI probability)
    quality_score = max(0, 10 - (ai_probability * 10))
    
    # Adjust quality based on positive signals
    if personal_count > 0:
        quality_score += 0.5
    if has_code:
        quality_score += 0.5
    if has_numbers:
        quality_score += 0.3
    if has_links:
        quality_score += 0.3
    
    quality_score = min(10, quality_score)
    
    return {
        'ai_probability': round(ai_probability, 2),
        'quality_score': round(quality_score, 1),
        'flags': flags,
        'recommendations': recommendations,
        'metrics': {
            'word_count': word_count,
            'slop_count': slop_count,
            'slop_phrases': slop_found,
            'em_dash_count': em_dash_count,
            'em_dash_ratio': round(em_dash_ratio, 2),
            'sentence_count': len(sentences),
            'avg_sentence_length': round(sum(sentence_lengths) / len(sentence_lengths), 1) if sentence_lengths else 0,
            'sentence_std_dev': round(std_dev, 1),
            'hedge_count': hedge_count,
            'hedge_ratio': round(hedge_ratio, 3),
            'vague_count': vague_count,
            'personal_count': personal_count,
            'has_code': has_code,
            'has_numbers': has_numbers,
            'has_links': has_links,
        }
    }

def print_detection_report(detection: Dict) -> None:
    """Print formatted detection report"""
    print("\n" + "="*60)
    print("AI SLOP DETECTION REPORT")
    print("="*60)
    
    print(f"\n📊 Overall Scores:")
    print(f"   AI Probability: {detection['ai_probability']*100:.0f}%")
    print(f"   Quality Score: {detection['quality_score']}/10")
    
    if detection['flags']:
        print(f"\n⚠️  Flags ({len(detection['flags'])}):")
        for flag in detection['flags']:
            print(f"   - {flag}")
    else:
        print(f"\n✅ No flags detected")
    
    if detection['recommendations']:
        print(f"\n💡 Recommendations:")
        for rec in detection['recommendations']:
            print(f"   - {rec}")
    
    metrics = detection['metrics']
    print(f"\n📈 Metrics:")
    print(f"   Words: {metrics['word_count']:,}")
    print(f"   Sentences: {metrics['sentence_count']}")
    print(f"   Avg Sentence Length: {metrics['avg_sentence_length']} words")
    print(f"   Sentence Variation (std dev): {metrics['sentence_std_dev']}")
    print(f"   AI Slop Phrases: {metrics['slop_count']}")
    print(f"   Em Dashes: {metrics['em_dash_count']} ({metrics['em_dash_ratio']} per 100 words)")
    print(f"   Personal Markers: {metrics['personal_count']}")
    print(f"   Has Code: {'✅' if metrics['has_code'] else '❌'}")
    print(f"   Has Numbers: {'✅' if metrics['has_numbers'] else '❌'}")
    print(f"   Has Links: {'✅' if metrics['has_links'] else '❌'}")
    
    if metrics['slop_phrases']:
        print(f"\n🚫 AI Slop Phrases Found:")
        for slop in metrics['slop_phrases']:
            print(f"   - '{slop['phrase']}' ({slop['count']}x)")
    
    print("\n" + "="*60)
    
    # Overall verdict
    if detection['quality_score'] >= 8:
        print("✅ PASS: Article meets quality standards")
    elif detection['quality_score'] >= 6:
        print("⚠️  REVIEW: Article needs minor improvements")
    else:
        print("❌ REJECT: Article needs major revision or regeneration")
    
    print("="*60 + "\n")
